_tutorials/2016-07-18-python-automation-netmiko-and-napalm.md

```diff
@@ -0,0 +1,226 @@
+---
+published: false
+date: '2016-07-18 23:11 +0300'
+title: 'Python automation: netmiko and napalm'
+author: Mike Korshunov
+tags:
+  - vagrant
+  - iosxr
+  - Python
+excerpt: Brief introduction to automation with Python
+---
+{% include toc icon="table" title="IOS-XR: Ansible and Vagrant" %}
+
+
+
+## Introduction
+
+At first, let's take a look at used tools:
+
+[Netmiko](https://github.com/ktbyers/netmiko) - multi-vendor ssh tool for device configuration 
+
+[NAPALM](https://github.com/napalm-automation/napalm)- tool, which provides standardized API for different platforms. 
+
+We will use old setup, which consist of devbox (Ubuntu instance) and rtr (IOS-XRv)  [link]({{ site.url }}/tutorials/IOSXR-Ansible).
+{: .notice--info}  
+
+We are interesting in IOS-XRv in particular. 
+
+### Module installation
+
+Let's install python modules on devbox:
+
+```shell
+sudo pip install netmiko
+
+sudo pip install napalm
+```
+
+As first step we can verify, that setup is successful. 
+Run python interpreter:
+
+```shell
+vagrant@vagrant-ubuntu-trusty-64:~$ python
+Python 2.7.6 (default, Jun 22 2015, 17:58:13)
+[GCC 4.8.2] on linux2
+Type "help", "copyright", "credits" or "license" for more information.
+>>> import netmiko
+>>> import napalm
+>>>
+
+```
+
+### Example via netmiko
+
+Create the first file and try to connect to the device. 
+
+```python
+
+from netmiko import ConnectHandler
+
+cisco_ios_xrv = {
+    'device_type': 'cisco_xr',
+    'ip':   '10.1.1.20',
+    'username': 'vagrant',
+    'password': 'vagrant',
+    'port' : 22,          # optional, defaults to 22
+    'secret': 'secret',     # optional, defaults to ''
+    'verbose': False,       # optional, defaults to False
+}
+
+net_connect = ConnectHandler(**cisco_ios_xrv)
+
+output = net_connect.send_command('show ip int brief')
+print(output)
+
+output = net_connect.send_config_set(['hostname my_sweet_rtr', 'commit'])
+print(output)
+
+```
+
```

Comment:
Show the effect of the send commands, Do a show running config using Netmiko if possible or by directly using a show running-config on XR CLI